keps/sig-auth/4193-tokenrequest-requester-info-enhancements/README.md

```diff
@@ -0,0 +1,623 @@
+### Extending TokenReview to allow cross-checking the embedded Node information with existing Node objects
+
+The TokenReview API will also be extended to check whether a JWT's embedded node information is still valid/current,
+and if not, will reject the review/indicate to the client that the token mismatches with the current state of Nodes.
+
+This will involve first checking whether the node object with the name given in the JWT still exists, and if it does,
+validating whether the UID of that node is equal to the UID embedded in the token.
+
+This means administrators can **delete node objects to invalidate all JWTs issued that embed that node's info**.
+
+To avoid unexpected breaking changes in behaviour, this behaviour will not only be protected by the feature flag whilst
+the feature is graduating to GA, but will also be gated behind an additional kube-apiserver flag which defaults to 'off'.
+
+### Including a UUID ([JTI](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.7)) on each issued JWT
+
+When a TokenRequest is being issued/fulfilled, we will modify the issuing code to also generate and embed a UUID which
+can be later used to trace the requests that a specific issued token has made to the apiserver via the audit log.
+
+This will require changing the JWT issuing code to actually generate this UUID, as well as extending the code around the
+audit log to have it record this information into audit entries.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Alice is building an identity issuance system that relies on verifying projected service account tokens to ensure that
+a pod exists, and is associated with a particular service account.
+
+Alice wants to prevent a malicious actor being able to replay this token from another host, so that she can be certain
+that the user requesting the identity document is also the same user that initially requested the token (as the identity
+used to request the token requires validations rooted in the cryptographic root of trust on the host, e.g. a TPM).
+
+This not only allows us to prevent replay attacks, but allows us to issue identity documents that we can 'root' in the
+node's own root of trust, the TPM (as the external service can now clearly validate that the user requesting the identity
+not only has a copy of that token, but also that it was the original entity to request that token).
+
+#### Story 2
+
+Bob is an administrator of a cluster and has noticed some strange request patterns from an unknown service account.
+
+Bob would like to understand who initially issued/authorised this token to be issued. To do so, Bob looks up the JTI
+of the token making the suspicious requests by looking inside the audit log entries for these suspect requests.
+
+This JTI is then used for a further audit log lookup - namely, looking for the TokenRequest `create` call which contains
+the audit annotation with key `authentication.kubernetes.io/jti` and the value set to that of the suspect token.
```

Comment:
```suggestion
the audit annotation with key `authentication.kubernetes.io/token-identifier` and the value set to that of the suspect token.
```

I'd make the annotation key implementation agnostic. JTI is what we're doing here, but maybe next week someone invents a gRPC Web Token :wink: or something and we actually have GTI.

Reply:
I like this suggestion.. have switched this to `token-identifier`, although speaking with @enj we have also noted we need to reach out to @tallclair as we may be better off making this a first class field rather than an annotation anyway: https://github.com/kubernetes/kubernetes/pull/118571#issuecomment-1662698887

Reply:
one possible step further... would we want to track credential identifiers from non-token auth methods as well, like an x509 client certificate (either issuer+serial or subject key identifier)? if so, something even more generic like credential-id might be a good idea

Reply:
X.509 is different enough from a token mechanism that I'd (soft) recommend a different set of annotations / fields.
We might record a hash of the entire certificate in one entry, the serial number from the cert in another, and the issuer's distinguished name in a third.